### PR TITLE
Better support for deploying to non-home tenant

### PR DESCRIPTION
**Pull request by pamelafox**
## Purpose

I've been deploying to my non-home tenant this week, and it's been tricky since DefaultAzureCredential doesn't necessarily know which tenant to use. Its chain is:
1) AzureCliCredential - which can have a default tenant override, if explicitly set
2) AzureDeveloperCliCredential - doesn't have a default tenant override, but will obey tenant_id or AZURE_TENANT_ID in global env

So sometimes my DAC will work, if I've got my az or env setup right, but often it will break.
This change makes our code more explicitly pass in the tenant_id for local use, which should be set by the azd up process.
However, its slightly backward incompatible as we were bringing tenantId into main.parameters.json, which was overriding what would otherwise be a correctly output tenantId and setting it to an empty string. So some developers may need to re-run azd provision, or go into their .env and set AZURE_TENANT_ID="" there, to get local dev working again.

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

```
[X] Yes - See above
[ ] No
```

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

```
[ ] Yes
[X] No
```

## Type of change

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [x] The current tests all pass (`python -m pytest`).
- [x] I added tests that prove my fix is effective or that my feature works
- [ ] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [x] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

### CODE REVIEW

**Comment by pamelafox** [comment]
CCing everyone who I've discussed DefaultAzureCredential woes with :)

File: app/backend/app.py [hunk 1]
```diff
@@ -436,11 +440,13 @@ async def setup_clients():
...
 
-    # Use the current user identity to authenticate with Azure OpenAI, AI Search and Blob Storage (no secrets needed,
-    # just use 'az login' locally, and managed identity when deployed on Azure). If you need to use keys, use separate AzureKeyCredential instances with the
-    # keys for each service
-    # If you encounter a blocking error during a DefaultAzureCredential resolution, you can exclude the problematic credential by using a parameter (ex. exclude_shared_token_cache_credential=True)
-    azure_credential = DefaultAzureCredential(exclude_shared_token_cache_credential=True)
+    # Use the current user identity for keyless authentication to Azure services.
+    # This assumes you use 'azd auth login' locally, and managed identity when deployed on Azure.
+    # The managed identity is setup in the infra/ folder.
+    if os.getenv("WEBSITE_HOSTNAME"):
```
Comment (tonybaloney) [inline app/backend/app.py:446]:
```suggestion
    if os.getenv("WEBSITE_HOSTNAME"):  # Environment variable set on Azure Web Apps
```

File: app/backend/prepdocs.py [hunk 2]
```diff
@@ -381,11 +381,12 @@ async def main(strategy: Strategy, setup_index: bool = True):
...
-        AzureDeveloperCliCredential()
-        if args.tenantid is None
-        else AzureDeveloperCliCredential(tenant_id=args.tenantid, process_timeout=60)
-    )
+    if args.tenantid:
+        logger.info("Connecting to Azure services using the azd credential for tenant %s", args.tenantid)
+        azd_credential = AzureDeveloperCliCredential(tenant_id=args.tenantid, process_timeout=60)
+    else:
+        logger.info("Connecting to Azure services using the azd credential for home tenant")
+        azd_credential = AzureDeveloperCliCredential()
```
Comment (tonybaloney) [inline app/backend/prepdocs.py:389]: Why does this one not have the same timeout?

Comment (pamelafox) [reply app/backend/prepdocs.py:389]: I don't know! I thought about removing it for consistency but then thought I shouldnt change things I dont know the origin for. Maybe @mattgotteiner recalls a reason?

File: .github/workflows/azure-dev-validation.yaml [hunk 3]
```diff
@@ -22,7 +22,9 @@ jobs:
       - name: Build Bicep for linting
         uses: azure/CLI@v2
         with:
-          inlineScript: az config set bicep.use_binary_from_path=false && az bicep build -f infra/main.bicep --stdout
+          inlineScript: |
+            export DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1
```
Comment (pamelafox) [inline .github/workflows/azure-dev-validation.yaml:26]: Fix for AZ CLI regression-
https://github.com/Azure/azure-cli/issues/29828

File: app/backend/prepdocs.py (hunk 2 above)
Comment (pamelafox) [reply app/backend/prepdocs.py:389]: Maybe we should put the timeout in both, just to give the credential more time to come back locally.

**Review by mattgotteiner — APPROVED** [review]
It's unfortunate DefaultAzureCredential didn't work here

File: app/backend/prepdocs.py (hunk 2 above)
Comment (mattgotteiner) [reply app/backend/prepdocs.py]: I don't think adding the timeout will have much issue. Sure, we can add it to both

**Comment by pamelafox** [comment]
@mattgotteiner  It works only if you're also logged into Azure CLI and have explicitly set your tenant's subscription to the one that was used in the app. So I think it's safer to go with the more explicit credential. We'll see if we get more/less issues with this sort of approach.

**Pull request merged**